ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/state/beaconstate/versions/eip4844/BeaconStateSchemaEip4844.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright ConsenSys Software Inc., 2022
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.eip4844;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.bellatrix.BeaconStateSchemaBellatrix.LATEST_EXECUTION_PAYLOAD_HEADER_FIELD_INDEX;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import tech.pegasys.teku.infrastructure.ssz.primitive.SszByte;
+import tech.pegasys.teku.infrastructure.ssz.schema.collections.SszPrimitiveListSchema;
+import tech.pegasys.teku.infrastructure.ssz.schema.collections.SszUInt64ListSchema;
+import tech.pegasys.teku.infrastructure.ssz.sos.SszField;
+import tech.pegasys.teku.infrastructure.ssz.tree.TreeNode;
+import tech.pegasys.teku.spec.config.SpecConfig;
+import tech.pegasys.teku.spec.config.SpecConfigEip4844;
+import tech.pegasys.teku.spec.datastructures.execution.versions.eip4844.ExecutionPayloadHeaderSchemaEip4844;
+import tech.pegasys.teku.spec.datastructures.state.SyncCommittee;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconStateSchema;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.common.AbstractBeaconStateSchema;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.common.BeaconStateFields;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.altair.BeaconStateSchemaAltair;
+
+public class BeaconStateSchemaEip4844
+    extends AbstractBeaconStateSchema<BeaconStateEip4844, MutableBeaconStateEip4844> {
+
+  @VisibleForTesting
+  BeaconStateSchemaEip4844(final SpecConfig specConfig) {
+    super("BeaconStateEip4844", getUniqueFields(specConfig), specConfig);
+  }
+
+  private static List<SszField> getUniqueFields(final SpecConfig specConfig) {
+    final SszField latestExecutionPayloadHeaderField =
+        new SszField(
+            LATEST_EXECUTION_PAYLOAD_HEADER_FIELD_INDEX,
+            BeaconStateFields.LATEST_EXECUTION_PAYLOAD_HEADER,
+            () -> new ExecutionPayloadHeaderSchemaEip4844(SpecConfigEip4844.required(specConfig)));
+    return Stream.concat(
+            BeaconStateSchemaAltair.getUniqueFields(specConfig).stream(),
+            Stream.of(latestExecutionPayloadHeaderField))
+        .collect(Collectors.toList());
```

Comment:
Won't this will be missing all the new fields added in Capella? It may be worth just getting all the capella fields and then replacing the latest execution payload header field. Would probably be quite useful to have a utility method around that can do that replacement.

Reply:
Looks like I copied it from Capella and it's the same bug here. Looking for a way to cover it in a test

Reply:
I need SpecLogic for tests, postpone

Reply:
Oh, it's not bug, @tbenr helped me to figure out

Reply:
I'd prefer to introduce an override pattern like this (this is for capella). To always build on top of the previous and just apply the differences.

```java
  private static List<SszField> getUniqueFields(final SpecConfig specConfig) {
    final SszField nextWithdrawalIndexField =
        new SszField(
            NEXT_WITHDRAWAL_INDEX,
            BeaconStateFields.NEXT_WITHDRAWAL_INDEX,
            () -> SszPrimitiveSchemas.UINT64_SCHEMA);
    final SszField nextWithdrawalValidatorIndexField =
        new SszField(
            NEXT_WITHDRAWAL_VALIDATOR_INDEX,
            BeaconStateFields.NEXT_WITHDRAWAL_VALIDATOR_INDEX,
            () -> SszPrimitiveSchemas.UINT64_SCHEMA);

    final Stream<SszField> overriddenBellatrixFields =
        BeaconStateSchemaBellatrix.getUniqueFields(specConfig).stream()
            .map(
                sszField -> {
                  if (sszField.getIndex() == LATEST_EXECUTION_PAYLOAD_HEADER_FIELD_INDEX) {
                    return new SszField(
                        LATEST_EXECUTION_PAYLOAD_HEADER_FIELD_INDEX,
                        BeaconStateFields.LATEST_EXECUTION_PAYLOAD_HEADER,
                        () ->
                            new ExecutionPayloadHeaderSchemaCapella(
                                SpecConfigCapella.required(specConfig)));
                  } else {
                    return sszField;
                  }
                });
    return Stream.concat(
            overriddenBellatrixFields,
            Stream.of(nextWithdrawalIndexField, nextWithdrawalValidatorIndexField))
        .collect(Collectors.toList());
  }
```

Reply:
otherwise we could end up being forced to always build on top of Altair and add keep adding everything from there for each additional fork

Reply:
And we can filter out fields when we will start deleting stuff (if we ever will)

Reply:
What about this? needs tests and refactoring, just to show an idea https://github.com/ConsenSys/teku/pull/6441/commits/9052835b49690342efd5c0d8bb285bf08100fe81

Reply:
decided to left it for a future PR

Reply:
Yeah it's worth nothing that Capella didn't inherit any of the fields from Bellatrix - the execution payload is a different schema in Capella so needed to be replaced.  4844 is the first point where we're wanting to inherit fields unchanged that aren't common to all forks.